KinesisIntegrationTest.restart_dynamicStream_graceful:359 [HZ-1541]

Got this test failure in [my PR](https://github.com/hazelcast/hazelcast/pull/22134#issuecomment-1238139461), and it's unrelated to PR changes.

```
[ERROR]   KinesisIntegrationTest.restart_dynamicStream_graceful:359->restart_dynamicStream:397->assertMessages:580->AbstractKinesisTest.assertMessages:159->HazelcastTestSupport.assertTrueEventually:1338->HazelcastTestSupport.assertTrueEventually:1236->AbstractKinesisTest.lambda$assertMessages$3:160 
Key sets differ! expected:<[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249]> but was:<[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 59, 60, 62, 63, 64, 65, 66, 69, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 83, 85, 87, 88, 91, 94, 96, 97, 98, 99, 100, 101, 102, 104, 106, 109, 111, 115, 116, 117, 119, 120, 121, 123, 124, 125, 126, 127, 128, 129, 131, 134, 136, 137, 138, 139, 140, 141, 144, 145, 147, 148, 149, 150, 152, 153, 154, 155, 156, 158, 159, 160, 161, 163, 164, 167, 168, 169, 170, 171, 172, 173, 174, 176, 177, 180, 181, 182, 183, 185, 187, 190, 191, 193, 194, 195, 196, 198, 200, 203, 204, 205, 207, 208, 209, 211, 212, 214, 215, 216, 218, 219, 220, 221, 222, 223, 224, 225, 230, 231, 232, 233, 234, 235, 236, 237, 238, 240, 241, 242, 243, 245, 246, 247, 249]>
```

This error was found in the logs: 
```
2022-09-26 00:43:01,359 [ INFO] [hz.blissful_grothendieck.jet.cooperative.thread-1] [c.h.j.i.e.TaskletExecutionService]: [127.0.0.1]:5702 [dev] [5.2.0-SNAPSHOT] Exception in ProcessorTasklet{08b2-0ea9-5140-0001/map#1}
java.lang.AssertionError: null
	at com.hazelcast.jet.impl.execution.KeyedWatermarkCoalescer.observeWm(KeyedWatermarkCoalescer.java:106) ~[hazelcast-5.2.0-SNAPSHOT.jar:5.2.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.fillInbox(ProcessorTasklet.java:590) ~[hazelcast-5.2.0-SNAPSHOT.jar:5.2.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.processInbox(ProcessorTasklet.java:484) ~[hazelcast-5.2.0-SNAPSHOT.jar:5.2.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:341) ~[hazelcast-5.2.0-SNAPSHOT.jar:5.2.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:336) ~[hazelcast-5.2.0-SNAPSHOT.jar:5.2.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:328) ~[hazelcast-5.2.0-SNAPSHOT.jar:5.2.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.call(ProcessorTasklet.java:291) ~[hazelcast-5.2.0-SNAPSHOT.jar:5.2.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:404) ~[hazelcast-5.2.0-SNAPSHOT.jar:5.2.0-SNAPSHOT]
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:895) ~[?:1.8.0_311]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:369) ~[hazelcast-5.2.0-SNAPSHOT.jar:5.2.0-SNAPSHOT]
	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_311] (edited)
```
credits: @orcunc

--- a/hazelcast/src/main/java/com/hazelcast/jet/impl/execution/KeyedWatermarkCoalescer.java
+++ b/hazelcast/src/main/java/com/hazelcast/jet/impl/execution/KeyedWatermarkCoalescer.java
@@ -85,6 +85,7 @@ public class KeyedWatermarkCoalescer {
         for (WatermarkCoalescer c : coalescers.values()) {
             c.observeEvent(queueIndex);
         }
+        idleQueues[queueIndex] = false;
     }
 
     public List<Watermark> observeWm(int queueIndex, Watermark watermark) {
@@ -92,8 +93,10 @@ public class KeyedWatermarkCoalescer {
             idleQueues[queueIndex] = true;
             boolean allIdle = true;
             // we need to track idle queues for the case when there's no coalescer yet
-            for (boolean idleQueue : idleQueues) {
-                allIdle &= idleQueue;
+            for (int i = 0; i < idleQueues.length; i++) {
+                if (!doneQueues.contains(i)) {
+                    allIdle &= idleQueues[i];
+                }
             }
 
             List<Watermark> watermarks = new ArrayList<>();

--- a/hazelcast/src/test/java/com/hazelcast/jet/impl/execution/KeyedWatermarkCoalescerTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/jet/impl/execution/KeyedWatermarkCoalescerTest.java
@@ -88,4 +88,17 @@ public class KeyedWatermarkCoalescerTest extends JetTestSupport {
         assertEquals(emptyList(), kwc.observeWm(0, IDLE_MESSAGE));
         assertEquals(asList(wm(42), IDLE_MESSAGE), kwc.observeWm(1, IDLE_MESSAGE));
     }
+
+    @Test
+    public void test_oneQueueDoneOtherIdle_then_allIdle() {
+        assertEquals(emptyList(), kwc.queueDone(0));
+        assertEquals(singletonList(IDLE_MESSAGE), kwc.observeWm(1, IDLE_MESSAGE));
+    }
+
+    @Test
+    public void test_q1IdleThenActiveThenQ2Idle_then_notAllIdle() {
+        assertEquals(emptyList(), kwc.observeWm(0, IDLE_MESSAGE));
+        kwc.observeEvent(0);
+        assertEquals(emptyList(), kwc.observeWm(1, IDLE_MESSAGE));
+    }
 }

--- a/hazelcast/src/test/java/com/hazelcast/jet/impl/execution/WatermarkCoalescerTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/jet/impl/execution/WatermarkCoalescerTest.java
@@ -186,6 +186,7 @@ public class WatermarkCoalescerTest {
     public void when_allDone_then_noMaxValueEmitted() {
         assertEquals(Long.MIN_VALUE, wc.queueDone(0));
         assertEquals(Long.MIN_VALUE, wc.queueDone(1));
+        assertFalse(wc.idleMessagePending());
     }
 
     @Test

